sw/device/lib/runtime/epmp.h

```diff
@@ -0,0 +1,550 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_RUNTIME_EPMP_H_
+#define OPENTITAN_SW_DEVICE_LIB_RUNTIME_EPMP_H_
+
+#include <stdint.h>
+
+/**
+ * RISC-V Enhanced Physical Memory Protection (EPMP).
+ *
+ * Specifications:
+ *  - PMP Enhancements for memory access and execution prevention on Machine
+ *    mode
+ *    https://docs.google.com/document/d/1Mh_aiHYxemL0umN3GTTw8vsbmzHZ_nxZXgjgOUzbvc8
+ *  - RISC-V Privileged Specfication (v. 20190608)
+ *    https://github.com/riscv/riscv-isa-manual/releases/download/Ratified-IMFDQC-and-Priv-v1.11/riscv-privileged-20190608.pdf
+ *  - Ibex PMP
+ *    https://ibex-core.readthedocs.io/en/latest/03_reference/pmp.html
+ *
+ * Assumptions (should be initialized in assembly but can be verified using
+ * `epmp_check`):
+ *   * Machine Mode Whitelist Policy is enabled (mseccfg.MMWP = 1)
+ *   * Machine Mode Lockdown is disabled (mseccfg.MML = 0)
+ */
+
+/**
+ * EPMP constants.
+ */
+typedef enum {
+  kEpmpGranularity = 0,
+  kEpmpNumRegions = 16,
+} epmp_constants_t;
+
+/**
+ * EPMP entry permissions.
+ *
+ * Entries configured with locked permissions may only be modified
+ * if Rule Locking Bypass is set (mseccfg.RLB = 1).
+ *
+ * When Machine Mode Lockdown is disabled (mseccfg.MML = 0) the
+ * combination R=0 W=1 is reserved. This is the assumed state and so
+ * it is not possible to set these values. The combination R=1 W=1
+ * X=1 has also been reserved by this library and may not be used
+ *
+ * An entry may only be configured with unlocked permissions if the
+ * entry is also configured as OFF.
+ *
+ * Note: permissions may have different meanings when Machine Mode
+ * Lockdown (mseccfg.MML) is set.
+ */
+typedef enum {
+  /**
+   * Unlocked with R=0, W=0 and X=0.
+   *
+   * Note: full access (R=1, W=1 and X=1) in Machine Mode. Only use
+   * in disabled (OFF) entries.
+   */
+  kEpmpPermUnlocked,
+
+  /**
+   * Locked with R=0, W=0 and X=0.
+   */
+  kEpmpPermLockedNoAccess,
+
+  /**
+   * Locked with R=0, W=0 and X=1.
+   */
+  kEpmpPermLockedExecuteOnly,
+
+  /**
+   * Locked with R=1, W=0 and X=0.
+   */
+  kEpmpPermLockedReadOnly,
+
+  /**
+   * Locked with R=1, W=0 and X=1.
+   */
+  kEpmpPermLockedReadExecute,
+
+  /**
+   * Locked with R=1, W=1 and X=0.
+   */
+  kEpmpPermLockedReadWrite,
+} epmp_perm_t;
+
+/**
+ * EPMP region specification.
+ *
+ * Provides the start and end addresses of a particular region. These addresses
+ * are byte-aligned (i.e. they are like regular pointers rather than encoded
```

Comment:
> byte-aligned

Any reason why we should not require `uint32_t` alignment?  We can optimize hardware integrity at a `uint32_t` granularity better compared to byte-aligned access.

Reply:
Sorry I think my wording here may be confusing things. My intent with that statement was to convey that the user should provide pointer values (i.e. `region.start = (uintptr_t)__RAM_START`) rather than `pmpaddr` values (i.e. `region.start = (uintptr_t)__RAM_START >> 2`). Does that answer your question?

PMP entries must always be aligned to (at least) a 4 byte boundary.

Reply:
ack. sg